synapse/replication/tcp/streams/_base.py

```diff
@@ -489,6 +497,8 @@ def current_token(self, instance_name: str) -> Token:
         return self.store.get_cache_stream_token_for_writer(instance_name)
 
     def minimal_local_current_token(self) -> Token:
+        if self.store._cache_id_gen:
+            return self.store._cache_id_gen.get_minimal_local_current_token()
```

Comment:
`_cache_id_gen` is None on SQLite

Reply:
Why is the cache stream special?

Reply:
I think at the time we didn't have a good way of doing it on SQLite, or something?

Reply:
Sorry, I think I mean two questions:

- Why are we handling the caches stream separately to the others? Presumably b/c it is especially chatty and we want to limit the updates sent out?
- What's the difference between `self.store._cache_id_gen.get_minimal_local_current_token()` and `self.current_token(self.local_instance_name)`?

Reply:
> Why are we handling the caches stream separately to the others? Presumably b/c it is especially chatty and we want to ratelimit the updates sent out?

This seems to be the case, by my reading of https://github.com/matrix-org/synapse/pull/16557/files#diff-844ba8f7be8c32eb75cc8092e1c48528797f6a8e1eeada942724ba6f73923a9cR214-R218

Reply:
> Why are we handling the caches stream separately to the others? Presumably b/c it is especially chatty and we want to ratelimit the updates sent out?

Note that we're defining things here differently only because we have to handle the case of `_cache_id_gen` is `None`. Otherwise the two current token functions work the same way as other streams.

https://github.com/matrix-org/synapse/pull/16557#discussion_r1373294538 has some discussion around why we don't want to send POSITION and why its safe to do so for caches stream. Broadly yes, its chatty and we don't need to care about the "global" maximum current token.


> What's the difference between self.store._cache_id_gen.get_minimal_local_current_token() and self.current_token(self.local_instance_name)?

`self.current_token(..)` is broadly just `self.store._cache_id_gen.current_token(..)`, except it returns 0 if `_cache_id_gen` is None. So they map directly to the difference between the `current_token(..)` and minimal tokens on the ID generators

Reply:
Ahh, I'd missed that this was inside the CachesStream class, I thought this was base logic for all streams.

I guess the point is that if there's no `_cache_id_gen` then there are no other workers to worry about and so the distinction between minimum and current tokens is moot?

Reply:
Are there any other streams whose `minimal_local_current_token` impl we should sanity check?

Reply:
I mean more that e.g. the function bodies here don't call something with the word "minimal" in:

https://github.com/matrix-org/synapse/blob/8f35f8148e1a7ce3ac249e2d2052854409f2c0d6/synapse/replication/tcp/streams/federation.py#L74-L75
https://github.com/matrix-org/synapse/blob/8f35f8148e1a7ce3ac249e2d2052854409f2c0d6/synapse/replication/tcp/streams/_base.py#L389-L390
https://github.com/matrix-org/synapse/blob/8f35f8148e1a7ce3ac249e2d2052854409f2c0d6/synapse/replication/tcp/streams/_base.py#L343-L344

Reply:
Right, so yeah you're right that they're suboptimal implementations, but they are also valid implementations. We mostly care about the difference for caches as a) its high traffic, and b) we have an extra check for it that wants `minimal_local_current_token` to return the actual minimum